connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala

```diff
@@ -163,6 +169,31 @@ case class SessionHolder(userId: String, sessionId: String, session: SparkSessio
       }
     }
   }
+
+  /**
+   * Caches given DataFrame with the ID. The cache does not expire. The entry needs to be
+   * explicitly removed by the owners of the DataFrame once it is not needed.
+   */
+  private[connect] def cacheDataFrameById(dfId: String, df: DataFrame): Unit = {
+    if (dataFrameCache.putIfAbsent(dfId, df) != null) {
+      throw new IllegalArgumentException(s"A dataframe is already associated with id $dfId")
```

Comment:
Cannot you use the error framework for this?

Reply:
Any existing ones I can use? This should be rare since it would only be caused by our bug. This is not very user visible. 
I didn't see much of framework errors used in connect server yet.

Reply:
> Any existing ones I can use?

Add new error class to error-classes.json, and raise `SparkException`.

> it would only be caused by our bug

Then we should consider `SparkException.internalError`

> I didn't see much of framework errors used in connect server yet.

Should start doing that if we are going to transfer errors/exceptions from server to client in a consistent way.

Reply:
I can use `SparkException.internalError`. Seems like it is not used anywhere in Spark yet.

Reply:
Nice. Updated it to use `SparkException.internalError`.